frontends/common/constantFolding.cpp

```diff
@@ -76,27 +69,26 @@ const IR::Node* DoConstantFolding::postorder(IR::Declaration_Constant* d) {
             ::error("%1%: Cannot evaluate initializer for constant", d->initializer);
         return d;
     }
-
-    if (typesKnown) {
-        // If we typechecked we're safe
-        setConstant(d, init);
-    } else {
-        // In fact, this declaration may imply a cast, so the actual value of
-        // d is not init, but (d->type)init.  The typechecker inserts casts,
-        // but if we run this before typechecking we have to be more conservative.
+    if (!typesKnown) {
+        // This declaration may imply a cast, so the actual value of d
+        // is not init, but (d->type)init. The typechecker inserts
+        // casts, but if we run this before typechecking we have to be
+        // more conservative.
         if (init->is<IR::Constant>()) {
             auto cst = init->to<IR::Constant>();
             if (d->type->is<IR::Type_Bits>()) {
                 if (cst->type->is<IR::Type_InfInt>() ||
                     (cst->type->is<IR::Type_Bits>() &&
                      !(*d->type->to<IR::Type_Bits>() == *cst->type->to<IR::Type_Bits>())))
                     init = new IR::Constant(init->srcInfo, d->type, cst->value, cst->base);
-                setConstant(d, init);
             }
         }
+        if (init != d->initializer)
+            d = new IR::Declaration_Constant(d->srcInfo, d->name, d->annotations, d->type, init);
     }
-    if (init != d->initializer)
-        d = new IR::Declaration_Constant(d->srcInfo, d->name, d->annotations, d->type, init);
+    auto orig = getOriginal()->to<IR::Declaration_Constant>();
+    BUG_CHECK(orig, "getOriginal() did not return a Declaration_Constant");
+    constants.emplace(orig, init);
```

Comment:
These 3 lines could be replaced by

    constants.emplace(getOriginal<IR::Declaration_Constant>(), init);

-- the templated version of `getOriginal` does the cast and the BUG_CHECK.